File: src/core/emit/opcodes.ts
/**
 * The forty-one instructions of `compiled-program.md` section 4, and the two
 * facts an emitter has to get right about each one.
 *
 * The operand count is here so that a mistyped instruction is caught where it
 * is built rather than by an engine's verifier three systems away. The depth
 * change is here because section 3.5 check 5 requires the stack depth to be
 * computable by walking the list, to agree at every join and to be zero at
 * `HALT`; a compiler that cannot compute it cannot know it emitted a program
 * any engine will load, and this is the table that makes it computable.
 *
 * `ARRAY`, `CALL_LIB` and `CALL_FN` have a depth that depends on an operand, so
 * their entries carry the rule rather than a number.
 */

export const OPCODES = [
  'CONST',
  'DUP',
  'POP',
  'LOAD',
  'STORE',
  'CELL_INIT',
  'LOADC',
  'STOREC',
  'SLOAD',
  'SSTORE',
  'HIST',
  'HISTP',
  'ADD',
  'SUB',
  'MUL',
  'DIV',
  'MOD',
  'NEG',
  'LT',
  'LE',
  'GT',
  'GE',
  'EQ',
  'NE',
  'NOT',
  'AND',
  'OR',
  'AND_SHORT',
  'OR_SHORT',
  'JUMP',
  'JUMP_FALSE',
  'TICK',
  'FOR_INIT',
  'FOR_NEXT',
  'ARRAY',
  'ELEM',
  'CALL_LIB',
  'CALL_FN',
  'RET',
  'EMIT',
  'HALT',
] as const;

export type Opcode = (typeof OPCODES)[number];

interface Shape {
  /** How many operands the instruction takes, section 4's tables. */
  readonly operands: number;
  /**
   * The change in stack depth, or the operand position it is computed from.
   *
   * `ARRAY n` is `1 - n` and `CALL_LIB f n st` is `1 - n`, so both read their
   * count from an operand. `CALL_FN` reads it from the call site, which is not
   * an operand of the instruction, so it is the one entry the walker has to be
   * told about.
   */
  readonly depth: number | 'arrayCount' | 'callCount' | 'callSite';
}

const SHAPES: Readonly<Record<Opcode, Shape>> = {
  CONST: { operands: 1, depth: 1 },
  DUP: { operands: 0, depth: 1 },
  POP: { operands: 0, depth: -1 },
  LOAD: { operands: 1, depth: 1 },
  STORE: { operands: 1, depth: -1 },
  CELL_INIT: { operands: 2, depth: 0 },
  LOADC: { operands: 1, depth: 1 },
  STOREC: { operands: 1, depth: -1 },
  SLOAD: { operands: 1, depth: 1 },
  SSTORE: { operands: 1, depth: -1 },
  HIST: { operands: 1, depth: 0 },
  HISTP: { operands: 1, depth: 0 },
  ADD: { operands: 0, depth: -1 },
  SUB: { operands: 0, depth: -1 },
  MUL: { operands: 0, depth: -1 },
  DIV: { operands: 0, depth: -1 },
  MOD: { operands: 0, depth: -1 },
  NEG: { operands: 0, depth: 0 },
  LT: { operands: 0, depth: -1 },
  LE: { operands: 0, depth: -1 },
  GT: { operands: 0, depth: -1 },
  GE: { operands: 0, depth: -1 },
  EQ: { operands: 0, depth: -1 },
  NE: { operands: 0, depth: -1 },
  NOT: { operands: 0, depth: 0 },
  AND: { operands: 0, depth: -1 },
  OR: { operands: 0, depth: -1 },
  AND_SHORT: { operands: 1, depth: 0 },
  OR_SHORT: { operands: 1, depth: 0 },
  JUMP: { operands: 1, depth: 0 },
  JUMP_FALSE: { operands: 1, depth: -1 },
  TICK: { operands: 1, depth: 0 },
  FOR_INIT: { operands: 5, depth: -3 },
  FOR_NEXT: { operands: 5, depth: 0 },
  ARRAY: { operands: 1, depth: 'arrayCount' },
  ELEM: { operands: 0, depth: -1 },
  CALL_LIB: { operands: 3, depth: 'callCount' },
  CALL_FN: { operands: 1, depth: 'callSite' },
  RET: { operands: 0, depth: -1 },
  EMIT: { operands: 1, depth: -1 },
  HALT: { operands: 0, depth: 0 },
};

export function operandCount(opcode: Opcode): number {
  return SHAPES[opcode].operands;
}

/**
 * The stack depth change of one instruction.
 *
 * `argcOf` answers for a `CALL_FN`, because its count lives in the call site
 * table rather than in the instruction, and the walker is the only caller that
 * has that table in hand.
 */
export function depthChange(
  opcode: Opcode,
  operands: readonly number[],
  argcOf: (site: number) => number,
): number {
  const shape = SHAPES[opcode];
  if (shape.depth === 'arrayCount') return 1 - (operands[0] ?? 0);
  if (shape.depth === 'callCount') return 1 - (operands[1] ?? 0);
  if (shape.depth === 'callSite') return 1 - argcOf(operands[0] ?? 0);
  return shape.depth;
}

/** The operand positions that hold a jump target, for the verifier's check 3. */
const TARGETS: Partial<Record<Opcode, readonly number[]>> = {
  CELL_INIT: [1],
  AND_SHORT: [0],
  OR_SHORT: [0],
  JUMP: [0],
  JUMP_FALSE: [0],
  FOR_INIT: [4],
  FOR_NEXT: [4],
};

export function targetPositions(opcode: Opcode): readonly number[] {
  return TARGETS[opcode] ?? [];
}

export function isOpcode(name: string): name is Opcode {
  return (OPCODES as readonly string[]).includes(name);
}

File: engine/openscript/opcodes.py
"""The forty-one instructions, as a table rather than as forty-one functions.

``compiled-program.md`` section 4.13 prints the whole set with its operands and
its stack effect, and this is that table in the one form an engine needs it:
what each opcode's operands are called, how deep the stack goes after it, which
operand is a jump target, and which table each operand indexes into.

It is a table because verification needs the same facts the interpreter needs. A
dispatch written as forty-one functions with the operand counts implied by their
signatures would leave the verifier to state them a second time, and the two
would drift the day an instruction changed.

**The page is the source and this is a copy, so a test compares them.** That is
the standing habit of this repository: a figure the specification prints is
measured by a test rather than quoted by one. ``tests/test_opcodes.py`` reads
section 4.13 out of the document by pattern and holds every row here to it, so a
row the page gains, loses or changes fails this engine's tests naming the
opcode. Nothing here may be corrected without the page being read again.
"""

from typing import Mapping, Sequence, Tuple

#: Depth changes that are not a fixed number, written as the page writes them.
BY_COUNT = "1 - n"
BY_ARGC = "1 - argc"

#: Opcode, operand names in order, depth change, and the group the page files it under.
TABLE: Mapping[str, Tuple[Sequence[str], object, str]] = {
    "CONST": (("k",), 1, "Constants and stack"),
    "DUP": ((), 1, "Constants and stack"),
    "POP": ((), -1, "Constants and stack"),
    "LOAD": (("s",), 1, "Slots"),
    "STORE": (("s",), -1, "Slots"),
    "CELL_INIT": (("c", "t"), 0, "Cells"),
    "LOADC": (("c",), 1, "Cells"),
    "STOREC": (("c",), -1, "Cells"),
    "SLOAD": (("r",), 1, "Series"),
    "SSTORE": (("r",), -1, "Series"),
    "HIST": (("r",), 0, "Series"),
    "HISTP": (("p",), 0, "Series"),
    "ADD": ((), -1, "Arithmetic"),
    "SUB": ((), -1, "Arithmetic"),
    "MUL": ((), -1, "Arithmetic"),
    "DIV": ((), -1, "Arithmetic"),
    "MOD": ((), -1, "Arithmetic"),
    "NEG": ((), 0, "Arithmetic"),
    "LT": ((), -1, "Comparison"),
    "LE": ((), -1, "Comparison"),
    "GT": ((), -1, "Comparison"),
    "GE": ((), -1, "Comparison"),
    "EQ": ((), -1, "Comparison"),
    "NE": ((), -1, "Comparison"),
    "NOT": ((), 0, "Logic"),
    "AND": ((), -1, "Logic"),
    "OR": ((), -1, "Logic"),
    "AND_SHORT": (("t",), 0, "Logic"),
    "OR_SHORT": (("t",), 0, "Logic"),
    "JUMP": (("t",), 0, "Control"),
    "JUMP_FALSE": (("t",), -1, "Control"),
    "TICK": (("l",), 0, "Control"),
    "FOR_INIT": (("l", "s", "sLim", "sStep", "t"), -3, "Control"),
    "FOR_NEXT": (("l", "s", "sLim", "sStep", "t"), 0, "Control"),
    "ARRAY": (("n",), BY_COUNT, "Arrays"),
    "ELEM": ((), -1, "Arrays"),
    "CALL_LIB": (("f", "n", "st"), BY_COUNT, "Calls"),
    "CALL_FN": (("site",), BY_ARGC, "Calls"),
    "RET": ((), -1, "Calls"),
    "EMIT": (("ch",), -1, "Output"),
    "HALT": ((), 0, "Termination"),
}

#: Which operand of an instruction is a jump target, by position.
#:
#: Read off the operand names rather than written down again: ``t`` is the name
#: section 4's operand table gives a jump target and nothing else carries it.
TARGETS: Mapping[str, Sequence[int]] = {
    opcode: tuple(at for at, name in enumerate(names) if name == "t")
    for opcode, (names, _depth, _group) in TABLE.items()
}

#: Which table each operand indexes into, for the range half of check 1.
#:
#: ``target`` is an instruction index in the same list and ``count`` is a number
#: the instruction carries rather than an index into anything.
OPERAND_TABLES: Mapping[str, Sequence[str]] = {
    "CONST": ("consts",),
    "LOAD": ("slots",),
    "STORE": ("slots",),
    "CELL_INIT": ("cells", "target"),
    "LOADC": ("cells",),
    "STOREC": ("cells",),
    "SLOAD": ("registers",),
    "SSTORE": ("registers",),
    "HIST": ("registers",),
    "HISTP": ("bindings",),
    "AND_SHORT": ("target",),
    "OR_SHORT": ("target",),
    "JUMP": ("target",),
    "JUMP_FALSE": ("target",),
    "TICK": ("loops",),
    "FOR_INIT": ("loops", "slots", "slots", "slots", "target"),
    "FOR_NEXT": ("loops", "slots", "slots", "slots", "target"),
    "ARRAY": ("count",),
    "CALL_LIB": ("libFunctions", "count", "states"),
    "CALL_FN": ("callSites",),
    "EMIT": ("channels",),
}

#: What a reader is told a table is called when an operand is outside it.
TABLE_NAMES: Mapping[str, str] = {
    "slots": "the frame",
    "cells": "cells",
    "states": "states",
    "registers": "series",
    "channels": "channels",
    "libFunctions": "lib.functions",
    "callSites": "callSites",
    "loops": "loops",
    "consts": "consts",
    "bindings": "the call site's series bindings",
}

#: Check 8's program half: an instruction that needs a tag in ``requires``.
INSTRUCTION_TAGS: Mapping[str, str] = {
    "ARRAY": "arrays",
    "ELEM": "arrays",
    "CALL_FN": "functions",
    "TICK": "loops",
}


def is_opcode(name: object) -> bool:
    return isinstance(name, str) and name in TABLE


def operand_count(opcode: str) -> int:
    return len(TABLE[opcode][0])


def operand_names(opcode: str) -> Sequence[str]:
    return TABLE[opcode][0]


def depth_change(opcode: str, operands: Sequence[int], argc_of) -> int:
    """How far the stack moves, with the two formulas resolved.

    ``CALL_FN`` reads its call site's ``argc``, which is fixed at load, so the
    depth stays statically computable and the verifier can walk the list.
    """
    written = TABLE[opcode][1]
    if written == BY_COUNT:
        return 1 - int(operands[1] if opcode == "CALL_LIB" else operands[0])
    if written == BY_ARGC:
        return 1 - int(argc_of(int(operands[0])))
    return int(written)


def targets(opcode: str) -> Sequence[int]:
    return TARGETS.get(opcode, ())


COUNT: int = len(TABLE)


def tables_for(opcode: str) -> Sequence[str]:
    return OPERAND_TABLES.get(opcode, ())
